_posts/2019-01-25-building-apps-with-plug-router.md

```diff
@@ -0,0 +1,345 @@
+Let's open `lib/router_example/router.ex` and update our plugs to include `Plug.Parsers` with `Jason` as our decoder:
+
+```elixir
+plug Plug.Parsers, parsers: [:json],
+                   pass: ["text/*"],
+                   json_decoder: Jason
+plug :match
+plug :dispatch
+```
+
+That's _really_ all we need to do for JSON.
+If we want to keep things simple, we can leverage `Jason.encode/1` or `Jason.encode!/1` along with `send_resp/3` and be done:
+
+```elixir
+{:ok, json} = Jason.encode(result)
+send_resp(conn, 200, json)
+```
+
+Or if we want a little more polish we could make a `render_json/2`:
+
+```elixir
+defp render_json(%{status: status} = conn, data) do
+  body = Jason.encode!(data)
+  send_resp(conn, (status || 200), body)
+end
+```
+
+For the remainder of the post we'll use the `render_json/2` approach.
+
+### Defining routes
+
+We now have our router code in place, code to render templates, and code to render JSON, all that's left is for us to define our routes!
+
+We previously created two EEx templates, `index.html.eex` and `contact.html.eex`, so we'll create routes to handle those requests first.
+From our healthcheck endpoint we know the format we expect but we can use our new `render/3` functions as well as our stubbed data.
+
+If you looked a `index.html.eex` then you know our EEx expects a `portfolio` capture populated with a list of maps containing `:name` and `:image`, the format we conveniently defined in our stubbed module!
+Let's bring all the pieces together:
+
+```elixir
+get "/" do
+  render(conn, "index.html", portfolio: Stubs.portfolio_entries())
+end
+
+get "/contact" do
+  render(conn, "contact.html")
+end
+```
+
+We're getting there but we're not _quite_ done.
+We have to handle our contact form's AJAX requests.
+
+In the interest of staying focused we won't get sidetrack with validation today, instead we'll make use of our `Stubs.submit_contact/1` function.
+Let's create a new route with the `post/1` macro that uses the aforementioned function and sends a pleasant JSON message back using our `render_json/2` function:
+
+```elixir
+post "/contact" do
+  Stubs.submit_contact(conn.params)
+  render_json(conn, %{message: "Thank you! We will get back to you shortly."})
+end
+```
+
+That's it, we're done, right?!
+Well — we probably want to handle requests to routes we haven't defined yet.
+Let's do that next and then we can call it done.
+
+### Missing routes
+
+Handling missing routes is a straight forward thanks to Elixir's powerful pattern matching.
+With the `match/3` macro and `_` we can match on all requests.
+By placing this at the bottom of our router we ensure if a request has not previously been matched it will be caught and handled.
+
+For now we'll implement a simple message, similar to how we implemented our "/ping" endpoint with `send_resp/3`:
+
+```elixir
+defmodule RouterExample.Router do
+  use Plug.Router
+
+  ...
+
+  match _ do
```

Comment:
Good note. In future posts that may go into further developing more complex web stuff with these plumbing tools this might be something that is a great starting point of showing how Elixir offers powerful abstractions to handle something that other languages might have a lot more boilerplate to deal with. I'm all for subtly embedding the lessons of the power of PM.